Formalize some SourceFile parsing outputs

hamishknight opened the pull request:
Currently when parsing a `SourceFile`, the parser gets handed pointers so that it can write the interface hash and collected tokens directly into the file. It can also call `setSyntaxRoot` at the end of parsing to set the syntax tree.

In preparation for the removal of `performParseOnly`, this PR formalizes these values as outputs of `ParseSourceFileRequest`, ensuring that the file gets parsed when the interface hash, collected tokens, or syntax tree is queried.

In addition, it cleans up `SourceFile` a little by sinking a couple of parsing options into `SourceFile::ParsingFlags`.

hamishknight committed:
NFC: Rename shouldCollectToken -> shouldCollectTokens

And fix the name of the underlying vector.

hamishknight committed:
[Parse] Remove some unnecessary code

The `SaveAndRestore` is unnecessary as `Parser`'s
constructor already sets up the interface hash,
and the request covers the `FrontendStatsTracer`.

hamishknight committed:
Sink some parsing options into SourceFile::ParsingFlags

Sink the `BuildSyntaxTree` and
`CollectParsedTokens` bits into
`SourceFile::ParsingFlags`, with a static method
to get the parsing options from the lang opts.

Also add a parsing flag for enabling the interface
hash, which can be used instead of calling
`enableInterfaceHash`.

hamishknight committed:
NFC: Inline canBeParsedInFull

Now that SIL files no longer interleave parsing
with type-checking, the query doesn't make much
sense. Inline it into its only client,
`shouldBuildSyntaxTree`.

hamishknight committed:
NFC: Clean up TokenRecorder a little

Rename `Bag` to `Tokens`, and query the
SourceManager from the ASTContext instead of
storing it directly.

hamishknight committed:
Formalize some SourceFile parsing outputs

Currently when parsing a SourceFile, the parser
gets handed pointers so that it can write the
interface hash and collected tokens directly into
the file. It can also call `setSyntaxRoot` at
the end of parsing to set the syntax tree.

In preparation for the removal of
`performParseOnly`, this commit formalizes these
values as outputs of `ParseSourceFileRequest`,
ensuring that the file gets parsed when the
interface hash, collected tokens, or syntax tree
is queried.

hamishknight committed:
[Parse] Make sure we don't finalize a DelayedTokenReciever

This is temporarily swapped in as the token
receiver while backtracking, so make sure we don't
try to call `finalize` on it.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

CodaFi commented on include/swift/Parse/Parser.h:
```diff
@@ -416,6 +423,14 @@ class Parser {
+  /// Retrieve the token receiver from the parser once it has finished parsing.
+  std::unique_ptr<ConsumeTokenReceiver> takeTokenReceiver() {
+    assert(Tok.is(tok::eof) && "not done parsing yet");
+    auto *receiver = TokReceiver;
```
@rintaro Does it make sense to model `TokReceiver` as a `std::unique_ptr` as well? It would make the ownership transfer here a little cleaner to express.

hamishknight replied:
I was thinking about that too, unfortunately we currently swap in a pointer to a `DelayedTokenReceiver` on the stack while backtracking.

CodaFi replied:
😢

CodaFi commented on lib/AST/Decl.cpp:
```diff
@@ -44,7 +44,7 @@
 #include "swift/AST/TypeLoc.h"
 #include "swift/AST/SwiftNameTranslation.h"
-#include "swift/Parse/Lexer.h"
+#include "swift/Parse/Lexer.h" // FIXME: Bad dependency
```
What's up here?

hamishknight replied:
Just something I noticed while working on the PR, as I understand it, libAST shouldn't depend on libParse.

CodaFi replied:
Ah, I read the diff wrong. I thought you had *introduced* this file.

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please clean test macOS

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test Linux

hamishknight committed:
[Parse] Preserve original member hashing behaviour

A previous commit inadvertently changed the
logic such that the member hash of an extension
body would be set to a partial interface hash.
Luckily this shouldn't have caused any behavioural
change as the interface hash itself would have
been left unaffected.

This commit makes sure we preserve the original
behaviour where if we don't have the body tokens
hashed separately, we give the body hash a default
constructed MD5.

Noticed by inspection.

hamishknight committed:
[AST] Remove hasSyntaxRoot

Use `shouldBuildSyntaxTree` instead.

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/1307

@swift-ci please test macOS

hamishknight commented:
Merging so I can make further progress – @rintaro I'm more than happy to address any review feedback in a follow-up.
